src/services/kratos/auth-device-account.ts

```diff
@@ -0,0 +1,249 @@
+        createIdentityBody: identityBody,
+      })
+      console.log(identityResults.status)
+      const result = await loginDeviceAccount({ deviceId })
+      if (result instanceof KratosError) return result
+      const sessionToken = result.sessionToken
+      const kratosUserId = result.kratosUserId as UserId
+
+      const levelZeroAccountsConfig = getDefaultAccountsConfig()
+      levelZeroAccountsConfig.initialLevel = AccountLevel.Zero
+      const account = await createAccountForDeviceAccount({
+        userId: kratosUserId,
+        config: levelZeroAccountsConfig,
+        device: deviceId,
+      })
+      if (account instanceof Error) return account
+      return { sessionToken, kratosUserId }
+    } catch (err) {
+      if (err.message === "Request failed with status code 400") {
+        return new LikelyUserAlreadyExistError(err.message || err)
+      }
+
+      return new UnknownKratosError(err.message || err)
+    }
+  }
+
+  const upgradeToPhoneSchema = async ({
+    kratosUserId,
+    phone,
+  }: {
+    kratosUserId: UserId
+    phone: PhoneNumber
+  }) => {
+    let identity: Identity
+
+    try {
+      ;({ data: identity } = await kratosAdmin.getIdentity({ id: kratosUserId }))
+    } catch (err) {
+      if (err.message === "Request failed with status code 400") {
+        return new LikelyUserAlreadyExistError(err)
+      }
+
+      return new UnknownKratosError(err)
+    }
+
+    if (identity.schema_id !== "device_account_v0") {
+      return new IncompatibleSchemaUpgradeError()
+    }
+
+    // Save a history of the device to the kratos admin metadata
+    try {
+      // in kratos datetime format
+      const updated_at = getKratosDateNow()
+      if (identity.metadata_admin) {
+        // is array
+        if (Array.isArray(identity.metadata_admin)) {
+          identity.metadata_admin = [
+            ...identity.metadata_admin,
+            { device: identity.traits.device, updated_at },
+          ]
+        } else {
+          // is json object
+          identity.metadata_admin = [
+            identity.metadata_admin,
+            { device: identity.traits.device, updated_at },
+          ]
+        }
+      } else {
+        // new data
+        identity.metadata_admin = [{ device: identity.traits.device, updated_at }]
+      }
+    } catch (err) {
+      console.log(err)
+    }
+
+    if (identity.state === undefined)
+      throw new KratosError("state undefined, probably impossible state") // type issue
+
+    identity.traits = { phone }
+
+    const adminIdentity: UpdateIdentityBody = {
+      ...identity,
+      credentials: { password: { config: { password } } },
+      state: identity.state,
+      schema_id: "phone_no_password_v0",
+    }
+
+    const { data: newIdentity } = await kratosAdmin.updateIdentity({
+      id: kratosUserId,
+      updateIdentityBody: adminIdentity,
+    })
+
+    return toDomainIdentityPhone(newIdentity)
+  }
+
+  const verifyJwt = async (token: string) => {
```

Comment:
is that not an (better?) option to call Oathkeeper api (the internal API) to verify the JWT?

Reply:
i don't think using oathkeeper will work because this is only called when creating a new device account, the kratos account does yet exist, so we can't call `http://localhost:4456/decisions/graphql` with the MeQuery for example.

to make this maybe work, i would have to create a separate oathkeeper rule for a specific graphql or kratos router endpoint to simply pass back the verified jwt data. seems like a lot of extra hops and config when we can simply verify the jwt with the jwks.

Reply:
i figured it out using sendOathkeeperRequest